src/main/java/com/android/volley/cronet/CronetHttpStack.java

```diff
@@ -0,0 +1,284 @@
+package com.android.volley.cronet;
+
+import android.content.Context;
+import androidx.annotation.Nullable;
+import androidx.annotation.VisibleForTesting;
+import com.android.volley.AuthFailureError;
+import com.android.volley.Header;
+import com.android.volley.Request;
+import com.android.volley.toolbox.AsyncHttpStack;
+import com.android.volley.toolbox.ByteArrayPool;
+import com.android.volley.toolbox.HttpResponse;
+import com.android.volley.toolbox.PoolingByteArrayOutputStream;
+import com.android.volley.toolbox.UrlRewriter;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.channels.Channels;
+import java.nio.channels.WritableByteChannel;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.Executor;
+import java.util.concurrent.ExecutorService;
+import org.chromium.net.CronetEngine;
+import org.chromium.net.CronetException;
+import org.chromium.net.UploadDataProvider;
+import org.chromium.net.UploadDataProviders;
+import org.chromium.net.UrlRequest;
+import org.chromium.net.UrlRequest.Callback;
+import org.chromium.net.UrlResponseInfo;
+
+/**
+ * A {@link AsyncHttpStack} that's based on Cronet's fully asynchronous API for network requests.
+ */
+public class CronetHttpStack extends AsyncHttpStack {
+
+    private static final int DEFAULT_POOL_SIZE = 4096;
+    private final CronetEngine mCronetEngine;
+    private final Executor mCallbackExecutor;
+    private final ExecutorService mBlockingExecutor;
+    private final ByteArrayPool mPool;
+    private final UrlRewriter mUrlRewriter;
+
+    public CronetHttpStack(
```

Comment:
Let's also allow callers to provide their own CronetEngine, but add Javadoc so we can suggest reasonable parameters, like not enabling the HTTP cache since it's redundant with Volley's cache.

We can continue to have a constructor which doesn't take a CronetEngine and creates a reasonable default, though.

We might also want to consider just having one private master constructor and adding a Builder class to construct stacks. We haven't done this elsewhere in Volley yet, but given the number of parameters and constructors we're getting here, and since this is a brand new class anyway, I think it might be a reasonable option.

Reply:
I ended up adding a Builder class, and allowing the caller to set their own CronetEngine, but defaulting to CronetEngine.build(context).